types/link.go

```diff
@@ -0,0 +1,218 @@
+	var lt LinkDefinitionType
+
+	// if no type is specified, we assume that brief notation of a link definition is used.
+	if a.Type == "" {
+		lt = LinkTypeBrief
+		r.Type = string(LinkTypeBrief)
+	} else {
+		r.Type = a.Type
+
+		lt, err = ParseLinkType(a.Type)
+		if err != nil {
+			return err
+		}
+	}
+
+	switch lt {
+	case LinkTypeVEth:
+		var l struct {
+			// the Type field is injected artificially
+			// to allow strict yaml parsing to work.
+			Type        string `yaml:"type"`
+			RawVEthLink `yaml:",inline"`
+		}
+		err := unmarshal(&l)
+		if err != nil {
+			return err
+		}
+		r.LinkConfig = *l.RawVEthLink.ToLinkConfig()
+	case LinkTypeMgmtNet:
+		var l struct {
+			Type           string `yaml:"type"`
+			RawMgmtNetLink `yaml:",inline"`
+		}
+		err := unmarshal(&l)
+		if err != nil {
+			return err
+		}
+		r.LinkConfig = *l.RawMgmtNetLink.ToLinkConfig()
+	case LinkTypeHost:
+		var l struct {
+			Type        string `yaml:"type"`
+			RawHostLink `yaml:",inline"`
+		}
+		err := unmarshal(&l)
+		if err != nil {
+			return err
+		}
+		r.LinkConfig = *l.RawHostLink.ToLinkConfig()
+	case LinkTypeMacVLan:
+		var l struct {
+			Type           string `yaml:"type"`
+			RawMacVLanLink `yaml:",inline"`
+		}
+		err := unmarshal(&l)
+		if err != nil {
+			return err
+		}
+		r.LinkConfig = *l.RawMacVLanLink.ToLinkConfig()
+	case LinkTypeMacVTap:
+		var l struct {
+			Type           string `yaml:"type"`
+			RawMacVTapLink `yaml:",inline"`
+		}
+		err := unmarshal(&l)
+		if err != nil {
+			return err
+		}
+		r.LinkConfig = *l.RawMacVTapLink.ToLinkConfig()
+	case LinkTypeBrief:
+		// brief link's endpoint format
+		var l struct {
+			Type       string `yaml:"type"`
+			LinkConfig `yaml:",inline"`
+		}
+
+		err := unmarshal(&l)
+		if err != nil {
+			return err
+		}
+
+		r.Type = string(LinkTypeBrief)
+
+		lc, err := briefLinkConversion(&l.LinkConfig)
+		if err != nil {
+			return err
+		}
+
+		r.LinkConfig = *lc
+	default:
+		return fmt.Errorf("unknown link type %q", lt)
+	}
+
+	return nil
+}
+
+func briefLinkConversion(lc *LinkConfig) (*LinkConfig, error) {
+	// check two endpoints defined
+	if len(lc.Endpoints) != 2 {
+		return nil, fmt.Errorf("endpoint definition should consist of exactly 2 entries. %d provided", len(lc.Endpoints))
+	}
+	for _, v := range lc.Endpoints {
+		parts := strings.SplitN(v, ":", 2)
+		node := parts[0]
+
+		lt, err := ParseLinkType(node)
+		if err != nil {
+			// if the link type parsing from the node name did fail
+			// we continue, since the node name is not like veth or macvlan or the like
+			continue
+		}
+
+		// if the node name is equal to a LinkType, we check the Type and only allow the depricated format
+		// for old types. New once we force to use the new link format.
+		switch lt {
+		case LinkTypeMgmtNet:
+			continue
+		case LinkTypeHost:
+			continue
+		case LinkTypeMacVLan, LinkTypeMacVTap:
+			return nil, fmt.Errorf("Link type %q needs to be defined in new link format", string(lt))
+		}
+	}
+	return lc, nil
+}
```

Comment:
Hi @steiler 
I am close to finialize the review, and I am concerned about this piece.
I think we should allow using the brief format for macvlan links, there is nothing wrong with it. I am quite positive that users will in 90% cases use the brief link format, as the new link format is too verbose to type in.

With that in mind my view is that internally we should use the links foundation you created here, but we need to strive to keep the brief format supported. Otherwise people will complain that it is too hard to define topologies with clab.

So I wonder if you are ok with removing this function entirely and allowing us to keep brief link definition type.

Reply:
Yes, possible.
It was now a means to restrict the brief format to existing links (switch statement) but is also meant to, in one of the the next step perform the conversion from the brief format, which is now represented by the LinkConfig to the type specific implementation. Where LinkConfig is just the raw / input version of the brief format.

Reply:
removed conversion in [f2e3fd3](https://github.com/srl-labs/containerlab/pull/1450/commits/f2e3fd3416a1502f39070d2106e0fc4161bf0b63)

we will add it when it will convert to concrete link types